src/c1_koiter_steigmann/koiter_steigmann_curvable_bell_elements.h

```diff
@@ -824,68 +1427,287 @@ namespace oomph
+ /// boundary in the triangle mesh.
+ //==========================================================================
+ void KoiterSteigmannC1CurvableBellElement::fully_clamp_specified_boundary(
+  const unsigned& b,
+  const Vector<BoundaryConditionForC1PlateBending*>& boundary_values_pt,
+  TriangleMeshCurviLine* curviline_pt)
+ {
+
+  for (unsigned i_field=0;i_field<3;i_field++)
+   {
     const unsigned first_nodal_type_index =
-      this->first_nodal_type_index_for_field(i_field);
+     this->first_nodal_type_index_for_field(i_field);
+    
     const unsigned n_vertices = nu_node();
-
-#ifdef PARANOID
-    // Check that the dof number is a sensible value
-    unsigned n_type = nu_type_at_each_node();
-    if (k_type >= n_type)
-    {
-      throw OomphLibError(
-        "Foppl von Karman elements only have 6 Hermite deflection degrees\
-of freedom at internal points. They are {w ; w,x ; w,y ; w,xx ; w,xy ; w,yy}",
-        OOMPH_CURRENT_FUNCTION,
-        OOMPH_EXCEPTION_LOCATION);
-    }
+    
+    // These KS elements only have displacement dofs at vertices, we assume
+    // these are the first n_vertices nodes
+    for (unsigned n = 0; n < n_vertices; ++n)
+     {
+      // Get boundary node
+      BoundaryNode<Node>* nod_pt =
+       dynamic_cast<BoundaryNode<Node>*>(this->node_pt(n));
+      
+      // Check if it is on the boundary
+      if (nod_pt!=0)
+       {
+        if (nod_pt->is_on_boundary(b))
+         {
+#ifdef PARANOID
+          // We should only have one coordinate on this boundary
+          unsigned nzeta=nod_pt->ncoordinates_on_boundary(b);
+          if (nzeta!=1)
+           {
+            std::stringstream error_message;
+            error_message << "boundary coordinate must be 1D!"
+                          << std::endl;
+            throw OomphLibError(error_message.str(),
+                                OOMPH_CURRENT_FUNCTION,
+                                OOMPH_EXCEPTION_LOCATION);
+           }
 #endif
-
+          
+          // Get value itself from boundary condition object
+          Vector<double> zeta(nzeta);
+          nod_pt->get_coordinates_on_boundary(b,zeta);
+          
+          // Get the Jacobian from the GeomObject parametrising the boundary
+          GeomObject* geom_obj_pt=curviline_pt->geom_object_pt();
+          DenseMatrix<double> drdzeta(nzeta,2);
+          geom_obj_pt->dposition(zeta,drdzeta);
+          RankThreeTensor<double> d2rdzeta2(nzeta,2,2);
+          geom_obj_pt->d2position(zeta,d2rdzeta2);
+          double dtdzeta=sqrt(drdzeta(0,0)*drdzeta(0,0)+drdzeta(0,1)*drdzeta(0,1));
+          
+          // We have 6 Hermite deflection degrees
+          // of freedom points. They are {u ; u,x ; u,y ; u,xx ; u,xy ; u,yy}
+          // or their rotated counterparts {u ; u,n ; u,t ; u,nn ; u,nt ; u,tt}
+          unsigned nw_type=6; // hierher Aidan the function I copied this from (which I think wasn't ever used had nw_type_at_each_node();
```

Comment:
Yes, it's better to fetch the data from elsewhere than to hardcode it every time it is used. However, this function is naturally very "hardcoded", as it necessarily knows a lot about the dofs, so it doesn't matter too much.

Reply:
query removed